core/src/main/java/org/infinispan/scattered/impl/ScatteredVersionManagerImpl.java

```diff
@@ -267,15 +270,21 @@ public boolean isVersionActual(int segment, EntryVersion version) {
    }
 
    @Override
-   public void notifyKeyTransferFinished(int segment, boolean expectValues) {
-      SegmentState update = expectValues ? SegmentState.VALUE_TRANSFER : SegmentState.OWNED;
-      SegmentState previous;
-      do {
-         previous = segmentStates.get(segment);
-         // It is possible that the segment is not in KEY_TRANSFER state, but can be in BLOCKED states as well
-         // when the CONFIRM_REVOKED_SEGMENTS failed.
-         log.tracef("Finishing transfer for segment %d = %s -> %s", segment, previous, update);
-      } while (!segmentStates.compareAndSet(segment, previous, update));
+   public void notifyKeyTransferFinished(int segment, boolean expectValues, boolean cancelled) {
+      SegmentState update;
+      if (expectValues) {
+         update = SegmentState.VALUE_TRANSFER;
+      } else if (cancelled) {
+         // The transfer is cancelled when a newer topology is being installed.
+         update = SegmentState.NOT_OWNED;
+      } else {
+         update = SegmentState.OWNED;
+      }
+      // It is possible that the segment is not in KEY_TRANSFER state, but can be in BLOCKED states as well
+      // when the CONFIRM_REVOKED_SEGMENTS failed.
+      // Another possibility is when there is already another rebalance starting, then it might be NOT_OWNED
```

Comment:
I've already forgot the details, but it seems that `notifyKeyTransferFinished` can be called from `SSCI.onTaskCompletion` which is being executed when the keys transfer was just cancelled, because of the other rebalance starting. This other rebalance has reset the status of the segment. However, the transfer should be cancelled *before* the segment is unregistered... So I don't know if that can still happen in practice.

Anyway, it doesn't hurt because we're just allowing to do a no-op NOT_OWNED -> NOT_OWNED.